vault.go

```diff
@@ -0,0 +1,320 @@
+		"search_query":            []string{},
+		vs.client.Auth.ClientName: userEncodedSecret,
+	}
+	primaryItem := map[string]interface{}{
+		"id": itemName,
+	}
+
+	// Create client keys databag
+	if err := vs.client.DataBags.CreateItem(vaultName, keysItem); err != nil {
+		return nil, err
+	}
+
+	// Create primary databag
+	if err := vs.client.DataBags.CreateItem(vaultName, primaryItem); err != nil {
+		return nil, err
+	}
+
+	databagItem := DataBagItem(primaryItem)
+	databagKeysItem := DataBagItem(keysItem)
+
+	return &VaultItem{
+		DataBagItem:  &databagItem,
+		Name:         itemName,
+		Vault:        vaultName,
+		VaultService: vs,
+		Keys: &VaultItemKeys{
+			DataBagItem: &databagKeysItem,
+			Name:        keysItemName(itemName),
+		},
+	}, nil
+}
+
+// DeleteItem deletes an item from a data bag
+//   Chef API Docs: http://docs.getchef.com/api_chef_server.html#id22
+func (vs *VaultService) DeleteItem(vaultName, vaultItem string) error {
+	notFound := func(err error) error {
+		if err != nil {
+			if errRes, ok := err.(*ErrorResponse); ok {
+				if errRes.Response.StatusCode != 404 {
+					return err
+				}
+			} else {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// Delete item keys databag item
+	err := vs.client.DataBags.DeleteItem(vaultName, keysItemName(vaultItem))
+	if re := notFound(err); re != nil {
+		return re
+	}
+
+	// Delete primary databag item
+	err = vs.client.DataBags.DeleteItem(vaultName, vaultItem)
+	if re := notFound(err); re != nil {
+		return re
+	}
+
+	return nil
+}
+
+// GetItem fetches a VaultItem and loads client keys
+func (vs *VaultService) GetItem(vaultName, itemName string) (*VaultItem, error) {
+	databagItem, err := vs.client.DataBags.GetItem(vaultName, itemName)
+	if err != nil {
+		return nil, err
+	}
+
+	item := &VaultItem{
+		Name:         itemName,
+		VaultService: vs,
+		DataBagItem:  &databagItem,
+		Vault:        vaultName,
+	}
+
+	vaultKeys, err := item.loadKeys(vs.client, vaultName, itemName)
+	if err != nil {
+		return nil, err
+	}
+
+	item.Keys = vaultKeys
+
+	return item, nil
+}
+
+// UpdateItem sets the item data, encrypts with a shared key, and then encrypts the shared key with each authorized client key in the <item>_keys data bag
+func (vs *VaultService) UpdateItem(item *VaultItem, data map[string]interface{}) error {
+	itemData := map[string]interface{}{}
+	sharedSecret, err := item.sharedSecret()
+	if err != nil {
+		return err
+	}
+
+	for key, value := range data {
+		if key == "id" {
+			// do nothing
+		} else {
+			wrappedData, err := json.Marshal(map[string]interface{}{"json_wrapper": value})
+			if err != nil {
+				return err
+			}
+			encryptedValue, err := encryptItemValue(wrappedData, sharedSecret)
+			itemData[key] = encryptedValue
+		}
+	}
+	itemData["id"] = item.Name
+
+	err = vs.client.DataBags.UpdateItem(item.Vault, item.Name, itemData)
+	if err != nil {
+		return err
+	}
+
+	databagData := DataBagItem(itemData)
+
+	item.DataBagItem = &databagData
+
```

Comment:
We need to reencrypt the shared secret for the users  in admins and clients as is we can't specify admins or clients which makes the feature hard to use.  The search feature is also not implemented.  I'm looking at adding admins and clients.